lxd/main_waitready.go

```diff
@@ -38,67 +40,64 @@ func (c *cmdWaitready) Command() *cobra.Command {
 func (c *cmdWaitready) Run(cmd *cobra.Command, args []string) error {
-	finger := make(chan error, 1)
-	var errLast error
-	go func() {
-		for i := 0; ; i++ {
-			// Start logging only after the 10'th attempt (about 5
-			// seconds). Then after the 30'th attempt (about 15
-			// seconds), log only only one attempt every 10
-			// attempts (about 5 seconds), to avoid being too
-			// verbose.
-			doLog := false
-			if i > 10 {
-				doLog = i < 30 || ((i % 10) == 0)
-			}
+	ctx := context.Background() // Default to no timeout.
+	if c.flagTimeout > 0 {
+		var cancel context.CancelFunc
 
-			if doLog {
-				logger.Debugf("Connecting to LXD daemon (attempt %d)", i)
-			}
+		// Add 100ms longer than timeout to allow receipt of response message from server so we can
+		// differentiate between the server being unreachable or just not ready yet.
+		ctx, cancel = context.WithTimeout(ctx, (time.Duration(c.flagTimeout)*time.Second)+(time.Millisecond*100))
+		defer cancel()
+	}
 
-			d, err := lxd.ConnectLXDUnix("", &lxd.ConnectionArgs{
-				SkipGetServer: true,
-			})
-			if err != nil {
-				errLast = err
-				if doLog {
-					logger.Debugf("Failed connecting to LXD daemon (attempt %d): %v", i, err)
-				}
-
-				time.Sleep(time.Second)
-				continue
-			}
+	log := func(i int, format string, args ...any) {
+		// Start logging only after the 10'th attempt (about 5seconds).
+		// Then after the 30'th attempt (about 15 seconds), log only only one attempt every 10 attempts
+		// (about 5 seconds), to avoid being too verbose.
+		doLog := false
+		if i > 10 {
+			doLog = i < 30 || ((i % 10) == 0)
+		}
 
-			if doLog {
-				logger.Debugf("Checking if LXD daemon is ready (attempt %d)", i)
-			}
+		if doLog {
+			logger.Debugf(format, args...)
+		}
+	}
 
-			u := api.NewURL().Path("internal", "ready").WithQuery("wait", "1")
-			_, _, err = d.RawQuery(http.MethodGet, u.String(), nil, "")
-			if err != nil {
-				errLast = err
-				if doLog {
-					logger.Debugf("Failed to check if LXD daemon is ready (attempt %d): %v", i, err)
-				}
+	for i := 0; ; i++ {
+		if ctx.Err() != nil {
+			return fmt.Errorf("LXD daemon not reachable after %ds timeout", c.flagTimeout)
+		}
 
-				time.Sleep(time.Second)
-				continue
-			}
+		if i > 0 {
+			time.Sleep(time.Second)
+		}
+
+		log(i, "Connecting to LXD daemon (attempt %d)", i)
 
-			finger <- nil
-			return
+		var d lxd.InstanceServer
+		d, err := lxd.ConnectLXDUnixWithContext(ctx, "", &lxd.ConnectionArgs{
+			SkipGetServer: true,
+		})
+		if err != nil {
+			log(i, "Failed connecting to LXD daemon (attempt %d): %v", i, err)
+			continue
 		}
-	}()
 
-	if c.flagTimeout > 0 {
-		select {
-		case <-finger:
-		case <-time.After(time.Second * time.Duration(c.flagTimeout)):
-			return fmt.Errorf("LXD still not running after %ds timeout (%v)", c.flagTimeout, errLast)
+		log(i, "Checking if LXD daemon is ready (attempt %d)", i)
+
+		u := api.NewURL().Path("internal", "ready").WithQuery("timeout", strconv.Itoa(c.flagTimeout))
```

Comment:
For consistency, should `flagTimeout` be treated (and changed to) `uint64`?
```suggestion
		u := api.NewURL().Path("internal", "ready").WithQuery("timeout", strconv.FormatUint(c.flagTimeout, 10))
```

Reply:
Yep will change

Reply:
https://github.com/canonical/lxd/pull/15501